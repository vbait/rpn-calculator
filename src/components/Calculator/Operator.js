/**
 * @typedef {Object} Operator
 * @property {string} name Operator name
 * @property {function} fn Operator function
 * @property {string} [regex=`\\${name}`] Regex for validation
 * @property {number} [countOperands=2] Count of function arguments
 */

/**
Operator for RPN Calculator
@constructor
@param {string} name Operator name
@param {function} fn Operator function
@param {string} [regex=`\\${name}`] Regex for validation
@param {number} [countOperands=2] Count of function arguments
*/
function Operator(name, fn, regex, countOperands = 2) {
  this.name = name;
  this.regex = regex || `\\${name}`;
  this.countOperands = countOperands;
  this.fn = fn;
}

export default Operator;
