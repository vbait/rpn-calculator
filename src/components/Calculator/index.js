export { default } from './Calculator';
export { default as Operator } from './Operator';
