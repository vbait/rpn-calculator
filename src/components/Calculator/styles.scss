$button-width: 60px;
$button-margin: 3px;

.calculator {
  display: inline-block;
  .form-control {
    margin: $button-margin;
    width: calc(100% - #{$button-margin} * 2);
  }
  button {
    outline: none;
    width: $button-width;
    margin: $button-margin;
    &.double {
      width: calc(#{$button-width} * 2 + #{$button-margin} * 2);
    }
  }
  .filter {
    padding: $button-margin;
  }
  .display {
    position: relative;
    textarea {
      height: 50px;
      min-height: 50px;
    }
    .alert {
      position: absolute;
      top: 0;
      left: $button-margin;
      width: calc(100% - #{$button-margin} * 2);
    }
  }
  .calculator-container {
    display: flex;
    .numbers {
      width: calc(#{$button-width} * 3 + #{$button-margin} * 3 * 2);
    }
    .operations {
      max-height: 180px;
      display: flex;
      flex-flow: column wrap;
    }
  }
  .manipulate {
    display: flex;
    button {
      width: 100%;
      text-transform: uppercase;
    }
  }
  .reset {
    text-align: right;
  }
}
