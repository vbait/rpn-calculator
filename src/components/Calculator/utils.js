/**
 * Returns an array with arrays of the given size.
 *
 * @param myArray {Array} array to split
 * @param chunk_size {Integer} Size of every group
 */
function chunkArray(myArray, chunkSize) {
  let index = 0;
  const arrayLength = myArray.length;
  const tempArray = [];

  for (index = 0; index < arrayLength; index += chunkSize) {
    const myChunk = myArray.slice(index, index + chunkSize);
    // Do something if you want with the group
    tempArray.push(myChunk);
  }
  return tempArray;
}

/**
 * @typedef {Object} Operator
 * @property {string} name Operator name
 * @property {function} fn Operator function
 * @property {string} [regex=`\\${name}`] Regex for validation
 * @property {number} [countOperands=2] Count of function arguments
 */

/**
 * Returns an regular expression to validate input data for RPN Calculator.
 *
 * @param {Object.<string, import('./Operator').Operator>} operators Object with operators (Operator)
 * @param {string} separator Separator between operators and operands
 * @return {string}
 */
function createValidatorRegExp(operators, separator) {
  const operatorsRegex = Object.entries(operators)
    .map(([, { regex: r }]) => {
      return `${r}`;
    })
    .join('|');
  return new RegExp(
    `^((-?\\d+\\.?\\d*${separator})|(-?\\d*\\.?\\d+${separator})|((${operatorsRegex})${separator}))*((-?\\d+\\.?\\d*)|(-?\\d*\\.?\\d+)|(${operatorsRegex}))$`,
  );
}

export { chunkArray, createValidatorRegExp };

// ^((-?\d+\.?\d* )|(-?\d*\.?\d+ )|((sin|cos|\-|\*) ))*((-?\d+\.?\d*)|(-?\d*\.?\d+)|(sin|cos|\-|\*))$
