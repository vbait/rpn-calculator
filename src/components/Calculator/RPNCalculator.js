import { createValidatorRegExp } from './utils';

/**
 * @typedef {import('./Operator').Operator} Operator
 */

/**
 * @typedef {Object} SEPARATOR
 * @property {string} name
 * @property {string} regex
 */

/** @constant {SEPARATOR} */
const SEPARATOR = { name: ' ', regex: ' ' };

/**
RPN Calculator
@constructor
*/
function Calculator() {
  const stack = []; // stack to keep all numbers

  /**
   * @type {Object.<string, Operator>}
   */
  const operators = {}; // all operators mapped by name
  let regex = new RegExp(); // all operators mapped by name

  /**
   * Validate if operator is presented
   * @param {string} operator Name of Operator
   * @returns {boolean}
   */
  function isOperatorValid(operator) {
    return !!operators[operator];
  }

  /**
   * Validate if has enough operands for operator
   * @param {string} st Stack of numbers
   * @param {number} operands Count of operands
   * @returns {boolean}
   */
  function isStackValid(st, operands) {
    return st.length >= operands;
  }

  /**
   * Return list of operators names
   * @returns {Array.<string>}
   */
  this.operators = () => Object.keys(operators);

  /**
   * Return separator object
   * @returns {SEPARATOR}
   */
  this.separator = () => SEPARATOR;

  /**
   * Return list of operands
   * @returns {Array.<number>}
   */
  this.numbers = () => [...stack];

  /**
   * Add new operators
   * @param {...Operator} args
   */
  this.addOperators = (...args) => {
    args.forEach((operator) => {
      operators[operator.name] = operator;
    });
    regex = createValidatorRegExp(operators, SEPARATOR.regex);
    return this;
  };

  /**
   * Add new values to stack and eval if there are operations
   * @param {string} input String with operators and operands
   * @throws Will throw an error if the operators are not valid or operands are not Number.
   */
  this.push = (input) => {
    this.validate(input);
    const temp = [...stack];
    input.split(SEPARATOR.name).forEach((value) => {
      if (isOperatorValid(value)) {
        const { fn, countOperands } = operators[value];
        if (!isStackValid(temp, countOperands)) {
          throw Error('Too many operators.');
        }
        const args = temp.splice(-1 * countOperands, countOperands);
        const result = parseFloat(fn(...args));
        if (Number.isNaN(result) || result === Infinity) {
          throw Error(
            `Cannot use operator "${value}" to values ${args.join(' ')}.`,
          );
        }
        temp.push(result);
      } else if (Number.isNaN(value) || Number.isNaN(parseFloat(value))) {
        throw Error(`Value "${value}" is not a number.`);
      } else {
        temp.push(parseFloat(value));
      }
    });
    stack.splice(0, stack.length, ...temp);
  };

  /**
   * Validate string
   * @param {string} input String with operators and operands
   * @throws Will throw an error if the operators are not valid or operands are not Number.
   */
  this.validate = (input) => {
    if (!regex.test(input)) {
      throw Error('Input value is not valid.');
    }
  };

  /**
   * Reset calculator
   */
  this.reset = () => {
    stack.splice(0, stack.length);
  };
}

export default Calculator;

// ^((-?\d+\.?\d* )|(-?\d*\.?\d+ )|((sin|cos|\-|\*) ))*((-?\d+\.?\d*)|(-?\d*\.?\d+)|(sin|cos|\-|\*))$
