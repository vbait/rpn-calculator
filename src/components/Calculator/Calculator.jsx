/* eslint-disable react/no-array-index-key */
import React, {
  useState,
  useCallback,
  useRef,
  useEffect,
  useMemo,
} from 'react';
import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import Alert from 'react-bootstrap/Alert';
import { chunkArray } from './utils';
import RPNCalculator from './RPNCalculator';
import Operator from './Operator';
import { NUMBERS } from './constants';
import './styles.scss';

const calculator = new RPNCalculator();
calculator.addOperators(
  new Operator('+', (a, b) => a + b),
  new Operator('-', (a, b) => a - b),
  new Operator('*', (a, b) => a * b),
  new Operator('/', (a, b) => a / b),
);

const separator = calculator.separator();
const separatorReplaceRegex = new RegExp(`(${separator.regex}){2,}`, 'g');

function Calculator({ operators: additionalOperators }) {
  const inputRef = useRef();
  const [operators, setOperators] = useState(calculator.operators());
  const [numbers, setNumbers] = useState(calculator.numbers());
  const [value, setValue] = useState('');
  const [error, setError] = useState();
  const [readonly, setReadonly] = useState(false);

  const handleReadonly = useCallback(() => {
    setReadonly((v) => !v);
  }, [setReadonly]);

  const handleChange = useCallback(
    (v) => {
      const formatted = v.replace(separatorReplaceRegex, separator.name);
      setError();
      setValue(formatted);
      inputRef.current.focus();
    },
    [setValue, setError, inputRef],
  );

  const handleApply = useCallback(() => {
    if (!value) return;
    try {
      calculator.eval(value.trim());
      setValue('');
      setNumbers(calculator.numbers());
    } catch (err) {
      setError(err.message);
    }
  }, [setError, setValue, setNumbers, value]);

  const handleBackSpace = useCallback(() => {
    setError();
    setValue((state) => {
      return state.slice(0, -1);
    });
    inputRef.current.focus();
  }, [setError, setValue]);

  const handleClear = useCallback(() => {
    setValue('');
    setError();
  }, [setValue]);

  const handleReset = useCallback(() => {
    calculator.reset();
    setValue('');
    setError();
    setNumbers(calculator.numbers());
  }, [setValue]);

  const handleInputKeyUp = useCallback(
    (event) => {
      if (event.keyCode === 13) {
        event.preventDefault();
        handleApply();
      }
    },
    [handleApply],
  );

  useEffect(() => {
    const input = inputRef.current;
    input.addEventListener('keyup', handleInputKeyUp);
    return () => {
      input.removeEventListener('keyup', handleInputKeyUp);
    };
  }, [inputRef, handleInputKeyUp]);

  useEffect(() => {
    calculator.addOperators(...additionalOperators);
    setOperators(calculator.operators());
  }, [additionalOperators, setOperators]);

  const chunkedOperators = useMemo(() => {
    return chunkArray(operators, 4);
  }, [operators]);

  return (
    <div className="calculator">
      <div className="filter">
        <Form.Check
          type="checkbox"
          id="id_readonly"
          label="Readonly"
          checked={readonly}
          onChange={handleReadonly}
        />
      </div>
      <div className="display">
        <input
          type="text"
          value={value}
          onChange={(e) => handleChange(e.target.value)}
          className="form-control"
          readOnly={readonly}
          ref={inputRef}
        />
      </div>
      <div className="display">
        <textarea
          className="form-control"
          readOnly
          rows="2"
          disabled
          value={numbers.join(' ')}
        />
        {!!error && <Alert variant="danger">{error}</Alert>}
      </div>
      <div className="reset">
        <Button variant="danger" onClick={handleBackSpace} disabled={!value}>
          {'<-'}
        </Button>
        <Button variant="danger" onClick={handleClear} disabled={!value}>
          CE
        </Button>
        <Button
          variant="danger"
          onClick={handleReset}
          disabled={!numbers.length && !value}
        >
          C
        </Button>
      </div>
      <div className="calculator-container">
        <div className="numbers">
          {NUMBERS.map((v) => (
            <Button key={v} onClick={() => handleChange(value + v)}>
              {v}
            </Button>
          ))}
          <Button onClick={() => handleChange(`${value}0`)} className="double">
            0
          </Button>
          <Button onClick={() => handleChange(`${value}.`)}>.</Button>
        </div>
        {chunkedOperators.map((rows, index) => (
          <div key={index} className="operations">
            {rows.map((operator) => (
              <Button
                key={operator}
                onClick={() => handleChange(value + operator)}
                variant="secondary"
              >
                {operator}
              </Button>
            ))}
          </div>
        ))}
      </div>
      <div className="manipulate">
        <Button
          onClick={() => handleChange(value + separator.name)}
          disabled={!value}
          variant="info"
        >
          space
        </Button>
        <Button
          onClick={handleApply}
          disabled={!value || !!error}
          variant="info"
        >
          apply
        </Button>
      </div>
    </div>
  );
}

Calculator.defaultProps = {
  operators: [],
};

Calculator.propTypes = {
  operators: PropTypes.arrayOf(PropTypes.instanceOf(Operator)),
};

export default Calculator;
