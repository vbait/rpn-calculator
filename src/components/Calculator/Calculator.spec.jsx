import React from 'react';
import Calculator from './Calculator';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(<Calculator />).toJSON();
  expect(tree).toMatchSnapshot();
});
