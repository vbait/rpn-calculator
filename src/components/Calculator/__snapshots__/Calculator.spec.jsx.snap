// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="calculator"
>
  <div
    className="filter"
  >
    <div
      className="form-check"
    >
      <input
        checked={false}
        className="form-check-input"
        disabled={false}
        id="id_readonly"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label"
        htmlFor="id_readonly"
        title=""
      >
        Readonly
      </label>
    </div>
  </div>
  <div
    className="display"
  >
    <input
      className="form-control"
      onChange={[Function]}
      readOnly={false}
      type="text"
      value=""
    />
  </div>
  <div
    className="display"
  >
    <textarea
      className="form-control"
      disabled={true}
      readOnly={true}
      rows="2"
      value=""
    />
  </div>
  <div
    className="reset"
  >
    <button
      className="btn btn-danger"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      &lt;-
    </button>
    <button
      className="btn btn-danger"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      CE
    </button>
    <button
      className="btn btn-danger"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      C
    </button>
  </div>
  <div
    className="calculator-container"
  >
    <div
      className="numbers"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="double btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        .
      </button>
    </div>
    <div
      className="operations"
    >
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        +
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        -
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        *
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        /
      </button>
    </div>
  </div>
  <div
    className="manipulate"
  >
    <button
      className="btn btn-info"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      space
    </button>
    <button
      className="btn btn-info"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      apply
    </button>
  </div>
</div>
`;
